src/cdomains/apron/affineEqualityDomain.apron.ml

```diff
@@ -0,0 +1,747 @@
+  sig
+    include Matrix with type vec := V(A).t and type num := A.t
+  end
+
+module ListMatrix : AbstractMatrix =
+  functor (A: Arith) (V: AbstractVector) ->
+  struct
+    include ConvenienceOps(A)
+    module V = V(A)
+    type t = V.t list
+
+    let empty () = []
+
+    let show x =
+      List.fold_left (^) "" (List.map (V.show) x)
+
+    let add_empty_column m pos =
+      List.map (fun y -> V.insert_val pos (of_int 0) y) m
+
+    let append_row m row  =
+      List.append m [row]
+
+    let get_row m n =
+      List.nth m n
+
+    let num_rows m =
+      List.length m
+
+    let remove_row m n =
+      if n < 0 || n >= num_rows m then failwith "Entry does not exist"
+      else List.filteri (fun i x -> n <> i) m
+
+    let get_col m n =
+      V.of_list (List.map (fun x -> V.nth x n) m)
+
+    let append_matrices m1 m2  =
+      List.append m1 m2
+
+    let num_cols m =
+      match m with
+      | [] -> 0
+      | x :: xs -> V.length x
+
+    let equal m1 m2 =
+      List.equal V.equal m1 m2
+
+    let subtract_rows_c row1 row2 c =
+      V.map2 (fun x y -> x -: (y *: c)) row1 row2
+
+    let reduce_row_to_zero m row_n col_n =
+      let red_row = get_row m row_n in
+      let c = V.nth red_row col_n in
+      List.map (fun x -> subtract_rows_c x red_row ((V.nth x col_n) /: c)) m
+
+    let remove_col m col_n =
+      let del_col = List.map (fun x -> V.remove_val x col_n) in
+      match List.findi (fun i x -> V.nth x col_n <> of_int 0)  (List.rev m) with
+      | exception Not_found -> del_col m
+      | (i, _) -> let len_i = List.length m - (i + 1)
+        in del_col (remove_row (reduce_row_to_zero m len_i col_n) len_i)
+
+    (*Gauss-Jordan Elimination to get matrix in reduced row echelon form (rref) + deletion of zero rows. None matrix has no solution*)
+
+    let map2i f m v = List.map2i f m (V.to_list v)
+
+    let normalize t =
+      let exception NoSolution in
+      let rec create_rref new_t curr_row =
+        if curr_row >= num_rows new_t then new_t else
+          let row = List.nth new_t curr_row in
+          match V.findi (fun i -> fun x -> x <> of_int 0) row with
+          | exception Not_found -> create_rref (remove_row new_t curr_row) curr_row
+          | (i, p) -> if V.compare_length_with row (i + 1) <= 0 then raise NoSolution
+            else let p = V.map (fun x -> x /: p) row in
+              let col = get_col new_t i in
+              let res = map2i (fun row_i y z -> if row_i <> curr_row then subtract_rows_c y p z else p) new_t col
+              in create_rref res (curr_row + 1)
+      in
+      match create_rref t 0 with
+      | c -> let sort v1 v2 =
+               let f = V.findi (fun i x -> x =: of_int 1) in
+               let (i1, _), (i2, _) = (f v1), (f v2) in Int.compare i1 i2
+        in Some (List.sort sort c)
+      | exception NoSolution -> None
+
+    let find_opt = List.find_opt
+
+    let map2 f m v = List.map2 f m (V.to_list v)
+
+    let of_list l = l
+
+    let to_list m = m
```

Comment:
Might be worth pulling all of this out into a separate module, making the domain less huge here.

Reply:
Ok but do you perhaps have suggestions where I could move it? I could maybe create a new file `Matrix.ml` or something like that and place it inside _src/cdomains_

Reply:
The module is currently called `VectorMatrix.ml ` in src/cdomains. Might not be the ideal location and could of course be moved